pkg/lib/serverTime.js

```diff
@@ -321,13 +313,131 @@
+        const disabled_file = cockpit.file(chronyd_sources_disabled, { superuser: "try" });
+
+        function parse_servers(data) {
+            if (!data)
+                return [];
+            const servers = [];
+            data.split("\n").forEach(function(line) {
+                const parts = line.split(" ");
+                if (parts[0] == "server")
+                    servers.push(parts[1]);
+            });
+            return servers;
+        }
+
+        return enabled_file.read()
+                .then(data => {
+                    if (data) {
+                        return {
+                            supported: true,
+                            enabled: true,
+                            servers: parse_servers(data)
+                        };
+                    } else {
+                        return disabled_file.read()
+                                .then(data => {
+                                    return {
+                                        supported: true,
+                                        enabled: false,
+                                        servers: parse_servers(data)
+                                    };
+                                });
+                    }
+                });
+    }
+
+    function set_custom_ntp_chronyd(servers, enabled) {
+        const enabled_file = cockpit.file(chronyd_sources_enabled, { superuser: "try" });
+        const disabled_file = cockpit.file(chronyd_sources_disabled, { superuser: "try" });
+
+        const text = "# This file is automatically generated by Cockpit\n\n" + servers.map(s => `server ${s}\n`).join("");
+
+        function ensure_sourcedir() {
+            function add_sourcedir(data) {
+                const line = "sourcedir " + chronyd_sourcedir;
+                if (data.indexOf(line) == -1)
+                    data += "\n# Sources added by the Cockpit web console\n" + line + "\n";
+                return data;
+            }
+            return cockpit.file("/etc/chrony.conf", { superuser: "try" }).modify(add_sourcedir);
+        }
+
+        return cockpit.spawn(["mkdir", "-p", chronyd_sourcedir], { superuser: "try" })
+                .then(() => {
+                    if (enabled)
+                        enabled_file.replace(text).then(() => disabled_file.replace(null)).then(ensure_sourcedir);
+                    else
+                        disabled_file.replace(text).then(() => enabled_file.replace(null));
+                });
+    }
+
+    self.get_custom_ntp = function () {
+        /* We support editing the configuration of systemd-timesyncd
+         * or chronyd, by dropping a config file snippet that we
+         * assume full control over.
+         */
+
+        const backend = get_timesync_backend();
+        if (backend == "timesyncd")
+            return get_custom_ntp_timesyncd();
+        else if (backend == "chronyd")
+            return get_custom_ntp_chronyd();
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17992-20230110-094617-6e1ebb8f-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)